### ISSUE TITLE
Simple config to set colors

### ISSUE THREAD

**Issue by robertmuehsig** [dotnet/docfx#1386]
I'm new to docFx, but I wonder if it would be possible to introduce a new set of color options to change the blue-ish (links) and green-ish (the menu seems to use a different color set) and header/footer background colors. 

![image](https://cloud.githubusercontent.com/assets/756703/23578576/4d5f800a-00da-11e7-8e22-6f7c6037d5a5.png)

I know I could replace the whole template, but this seems to go too far just to ajust some colors. Or is there a way to overwrite these with a simple css style?

**Comment by LordZoltan** [comment]
# TL;DR

I'm posting this as much so that other people can get a feel for how they might go about 'replacing the whole template', as you mentioned, in such a way that doesn't really require a lot of work or, crucially a lot of maintenance, and which will give you ultimate control.

There just seems to be a dearth of examples out there for working with docfx - and so I hope you don't mind me partially hijacking your issue temporarily to provide some help to one and all :)

What I'm posting here is how to take almost full control of how your site looks by replacing the backbone of its styles in the default theme: bootstrap.  It requires use of gulp or some other design/build-time automation to rebundle the `docfx.vendor.css` file, and a custom template.  It's not the only way to do it, certainly, but it's pretty powerful.

# Customising with gulp/less

If you're willing to go on a bit of a magical mystery tour of both docfx, and my mind(!), then you can do what I've done for the [docs site for my .Net Core-compatible IOC Container, 'Rezolver'](http://rezolver.co.uk), which overrides the default template to replace the `docfx.vendor.*` files with ones which contains a custom build of bootstrap, plus everything else they normally would, which are, in order:

- `docfx.vendor.css`
  - bootstrap 
  - 'github-gist' theme for highlightjs
- `docfx.vendor.js`
  - jquery
  - bootstrap
  - highlightjs
  - lunr
  - js-url
  - twbs-pagination
  - mark.js

I've [kinda documented the process at a high level here](http://rezolver.co.uk/developers/docs/miscellaneous/how-this-website-is-built.html) (which I will probably soon update now with what I've written here!).

> My site overrides more than just the styles (some templates, too) - but with just a custom-themed bootstrap you can do a lot.

You'll want to compare the [default theme's gulpfile](https://github.com/dotnet/docfx/blob/dev/src/docfx.website.themes/default/gulpfile.js) with mine (in the root of the site), specifically the ['vendor' object](https://github.com/ZolutionSoftware/Rezolver/blob/1.1.71702.100/doc/Rezolver.Documentation/gulpfile.js#L46) , the ['copy' task](https://github.com/ZolutionSoftware/Rezolver/blob/1.1.71702.100/doc/Rezolver.Documentation/gulpfile.js#L69),the ['less' task](https://github.com/ZolutionSoftware/Rezolver/blob/1.1.71702.100/doc/Rezolver.Documentation/gulpfile.js#L89), and the ['bundle' task](https://github.com/ZolutionSoftware/Rezolver/blob/1.1.71702.100/doc/Rezolver.Documentation/gulpfile.js#L111) all of which produce the combined files which eventually get pushed into styles folder that is generated by docfx when you build.

## A few notes

- You should make sure you have the correct packages in both your `bower.json` and `package.json` files for stuff like jquery/bootstrap etc, but also to support the gulp tasks you need
- My site's [`docfx.json`](https://github.com/ZolutionSoftware/Rezolver/blob/1.1.71702.100/doc/Rezolver.Documentation/docfx.json) is worth a peek
  - I've changed where the documentation md/yml files are picked up from so they're all under a subfolder to keep the project a bit cleaner.
  - I use a local *copy* of the default theme (in `~/_docfx_themes/default`) instead of using it by name - that's to protect myself from changes between versions.  None of the files in that folder, however, are modified after they're copied in.
  - I've added a few global properties which are picked up in some html template overrides I've added - for things like a github link (with icon, from fontawesome) and a version number which I include above the footer
- My less compilation compiles the [site.less](https://github.com/ZolutionSoftware/Rezolver/blob/1.1.71702.100/doc/Rezolver.Documentation/styles/site.less) file in `~/styles` into `site.css` - and *that's* the file that gets included in the `~/docfx/themes/rezolver/styles/docfx.vendor.css` by the `bundle` task linked above.  In this file I roll up bootstrap, fontawesome and a custom theme of highlightjs, so that it's automatically built into my template's version of `docfx.vendor.css`
- Colour customisations and other tweaks to the core bootstrap are done in [`~/styles/bootstrap-variables-rezolver.less`](https://github.com/ZolutionSoftware/Rezolver/blob/1.1.71702.100/doc/Rezolver.Documentation/styles/bootstrap-variables-rezolver.less)
- I alter the font path for both bootstrap and fontawesome so that both reside under the path `fonts/` relative to where the css will be loaded from
- I've also overridden some HTML templates, such as the `head.tmpl.partial` with hooks to inject browser icons and a google analytics script - but the style customisations you're after don't require them.
- My template also performs further style overrides in the [`~/_docfx_themes/rezolver/styles/main.less` file](https://github.com/ZolutionSoftware/Rezolver/blob/1.1.71702.100/doc/Rezolver.Documentation/_docfx_themes/rezolver/styles/main.less) (which overrides the default template's empty `main.css` file where you're usually encouraged to do your 'adjustments' once the site is built) to adjust how the breadcrumbs subnav looks, the side TOC, and to [label code blocks with the name of the code file](https://github.com/ZolutionSoftware/Rezolver/blob/1.1.71702.100/doc/Rezolver.Documentation/_docfx_themes/rezolver/styles/main.less#L79) among other things. *I keep meaning to add an issue here to ask for the code file name to be rendered for us as a link to the source code here on github if the code is a snippet!*
- This all sounds like a lot of work - but once you're up and running, the whole process is *so* simple.

My style customisations are probably not finished yet - I keep changing my mind about some of the colours - but at least now I know I'm just playing with LESS variables or targeting overrides at the few styles the default template's `docfx.css` file has in it (which my template does not override, and therefore gets used as-is).

## Handling updates in the docfx default template

If the team update their template, I will simply unpack it from the nuget package's executable again, copy the files into my site's local copy of the default template and do a diff to see if any of the changes affect the files that I've overriden.  If they do, I can simply manually merge those changes into my template overrides; and I can separately check their `gulpfile.js`

* * *

Anyway - as I say, this is just to help others who might be wondering how to go about working with the template stuff built into docfx - I'll disappear now and let them answer your actual issue!  Sorry for hijacking!

**Comment by vicancy** [comment]
Hi @robertmuehsig You can provide your `main.css` file in your template folder, e.g. `template`, to override the existing styles. And set your `docfx.json` as similar to https://github.com/dotnet/docfx/blob/dev/Documentation/docfx.json#L60. The mechanism of templates is that files in the latter folder (if any) overwrites files with the same name in the former ones.

**Comment by vicancy** [comment]
`main.css` file inside default template is an empty file designed for you to overwrite.

**Comment by robertmuehsig** [comment]
Ah, but I need to export the default template? Otherwise I can't "inject" the css, correct?

**Comment by vwxyzh** [comment]
@robertmuehsig You can create a template, it only contain files which you want to replace in default template (keep same path and same file name), then use `-t default,yours`, or set them in `docfx.json`.

**Comment by robertmuehsig** [comment]
Ah - I understand. I need to take a look at it. I close this for now - looks like a clever solution 👍
